### ISSUE TITLE
Where is the the Python (3) kernel?

### ISSUE THREAD

**Issue by lukejanicke** [nteract/archived-desktop-app#2347]
**nteract** is working for me with the Python kernel but I have a question.

I think I have `kernel.json` in two locations:

1. `/usr/local/share/jupyter/kernels/python3/kernel.json`
2. `/Users/lukejanicke/Library/Jupyter/kernels/python3/kernel.json`

Is that normal?

Here is my installation procedure.

1. I installed Python 3 with Homebrew.

```
brew install python3
```

2. [**nteract** website recommends](https://nteract.io/kernels/python) installing the Python kernel with the following commands.

```
python -m pip install ipykernel
python -m ipykernel install --user
```

The first command did not work for me.

```
~ $ python -m pip install ipykernel
/usr/bin/python: No module named pip
```

A few other attempts also failed.

```
~ $ python3 -m pip3 install ipykernel
/usr/local/opt/python3/bin/python3.6: No module named pip3
```

Until I found something that worked.

```
~ $ pip3 install ipykernel
Collecting ipykernel
. . .
Successfully installed appnope-0.1.0 decorator-4.1.2 ipykernel-4.7.0 ipython-6.2.1 ipython-genutils-0.2.0 jedi-0.11.1 jupyter-client-5.2.1 jupyter-core-4.4.0 parso-0.1.1 pexpect-4.3.1 pickleshare-0.7.4 prompt-toolkit-1.0.15 ptyprocess-0.5.2 pygments-2.2.0 python-dateutil-2.6.1 pyzmq-16.0.3 simplegeneric-0.8.1 six-1.11.0 tornado-4.5.3 traitlets-4.3.2 wcwidth-0.1.7
```

I later found that `python3 -m pip install ipykernel` also works.

At this point, `jupyter kernelspec list` gave the following output.

```
~ $ jupyter kernelspec list
Available kernels:
  python3    /usr/local/share/jupyter/kernels/python3
```

I ran the next command from the [**nteract** website](https://nteract.io/kernels/python), changing `python` to `python3`.

```
~$ python3 -m ipykernel install --user
Installed kernelspec python3 in /Users/lukejanicke/Library/Jupyter/kernels/python3
```

At this point, I tried to run **nteract** and the Python kernel was not starting.

I ran `jupyter kernelspec list` again and noticed a different result!

```
$ jupyter kernelspec list
Available kernels:
  python3    /Users/lukejanicke/Library/Jupyter/kernels/python3
```

This confused me, because I remember from the last time I did this that I had to go into `kernel.json` and change the first `argv` from `python` to `python3`.

But now I had two `kernel.json` files.

1. `/usr/local/share/jupyter/kernels/python3/kernel.json`
2. `/Users/lukejanicke/Library/Jupyter/kernels/python3/kernel.json`

`kernel.json` (2) already said `python3` and **nteract** was still not starting the Python kernel. So naturally, I looked at `kernel.json` (1). It still said `python`, so I changed that to `python3` as below.

```
{
 "argv": [
  "python3",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```

After this, the **nteract** started with the Python kernel and now seems to be working correctly.

Back to my question.

Why did `jupyter kernelspec list` give a different result between the following two commands?

```
pip3 install ipykernel
python3 -m ipykernel install --user
```

And why are there two `kernel.json` files?

**Comment by willingc** [comment]
Hi @lukejanicke. You've run into a common point of confusion when using multiple versions of Python. Let's see if I can give you a bit more insight.

Yes, there can be two or more kernel files (Python3, Python2, R, Julia, etc.). For example, on my system, I currently have two kernels installed (Python3 and Python2):
```
$ jupyter kernelspec list
Available kernels:
  python3    /anaconda3/share/jupyter/kernels/python3
  python2    /usr/local/share/jupyter/kernels/python2
```

### Finding your python versions

Unrelated to the kernels, I can find where I have Python installed by using the `which` command:
```
carol at cw-pro in ~
$ which python3
/anaconda3/bin/python3

carol at cw-pro in ~
$ which python2
/usr/local/bin/python2

carol at cw-pro in ~
$ which python
/anaconda3/bin/python
```
My system has the `/anaconda3/bin/python` symlinked (i.e. aliased) to `/anaconda3/bin/python3`.

### Understanding the nteract recommendation

The pattern that nteract recommends is a one that follows the Python Packaging good practice.
Looking at the recommendation that nteract gives:
```
python -m pip install ipykernel
python -m ipykernel install --user
```
In the case of my system, the command `python -m pip install ipykernel` would be the same as if I typed `/anaconda3/bin/python -m pip install ipykernel`. When the command executes python, it aliases to `/anaconda3/bin/python3`. By using `python -m pip`, you are essentially telling the system to use the version of pip installed with the particular version of Python that `which python` returns.

For example, `python2 -m pip install ipykernel` would use `/usr/local/bin/python2` to execute pip to install the kernel.

So to recap up to this point, there are different Python versions on my computer. I can find them with the `which` command. `python -m`, `python2 -m` and `python3 -m` are shortcuts to the different Pythons that are displayed when entering the `which` command.

Hope this helps a bit.

**Comment by rgbkrk** [comment]
This was so thorough @lukejanicke, thank you!

What this comes down to is that nteract runs as a native app and relies on whatever `PATH` is available to it from launchctl (the process manager the Mac uses for `.app`s). We do some shell initialization so that we can find your proper `python`, `node`, etc.

It could be that our shell invocation this isn't completely working or is a deeper sign of the default `python` on your system not being the `python3` executable, which means that the installed kernelspec is referencing the wrong `python` altogether. Since we don't launch in jupyter's python process, this doesn't get corrected by assuming that the python given is the `sys.prefix` of the current jupyter.

**Comment by lukejanicke** [comment]
As rgbkrk said, that was incredibly thorough. So I’m gonna leave it for the weekend to read and understand thoroughly but thanks in advance for responding so well. I look forward too reading and understanding.

**Comment by willingc** [comment]
Thanks @lukejanicke. Feel free to follow up with questions as well. The details of this stuff is confusing :-)
